### PR TITLE
[Fix-14625][dolphinscheduler-master] Submit the condition task behind the skip node

### LINKED ISSUE #14625: [Bug] [dolphinscheduler-master] In special cases, the conditions task cannot be triggered to execute
### Search before asking

- [X] I had searched in the [issues](https://github.com/apache/dolphinscheduler/issues?q=is%3Aissue) and found no similar issues.

### What happened

In special cases, the conditions task cannot be triggered to execute. As shown in the following figure, when s7 finish before s1, task c2 cannot be triggered, resulting in task s5 or s6 cannot be triggered. The condition for task c2 is or. 

(当任务s7先于任务s1完成且c2的条件定义为or时，s4被跳过导致condition类型任务c2无法被触发，导致后续的s5或者s6任务无法执行。s1至s6均为sleep 20s，s7为sleep 10s)

The workflow is shown below. Tasks s1 to s6 are shell tasks with sleep 20s. Task s7 is a shell type task and is sleep 10s. 
image
Condition task c1 is defined as shown in the following figure:
image
Condition task c2 is defined as shown in the following figure:
image

### What you expected to happen

Condition type task c2 is defined as OR, then it should be executed as long as one of the preceding tasks is completed.
（Condition类型任务c2被定义为OR，则只要前置任务有一个完成就应该被触发。）

### How to reproduce

Build the workflow shown above. As long as task s7 finishes before task s1, and the c2 task condition is or, the problem can be repeated
（如上文所示构建工作流，只要任务s7先于任务s1执行结束，且c2任务条件为or，就能复现c2无法被执行问题）

### Anything else

_No response_

### Version

3.1.x

### Are you willing to submit PR?

- [X] Yes I am willing to submit a PR!

### Code of Conduct

- [X] I agree to follow this project's [Code of Conduct](https://www.apache.org/foundation/policies/conduct)

### PR DESCRIPTION
**Pull request by shilei-ly**
Submit the condition task behind the skip node

## Purpose of the pull request
To fix a bug, see the issue 14625 for details。fix https://github.com/apache/dolphinscheduler/issues/14625

## Brief change log
Modification some method， eg: org.apache.dolphinscheduler.service.utils.DagHelper#parseConditionTask

## Verify this pull request

This pull request is code cleanup without any test coverage.

*(or)*

This pull request is already covered by existing tests, such as *(please describe tests)*.

(or)

This change added tests and can be verified as follows:
Modify org.apache.dolphinscheduler.service.utils.DagHelperTest#testConditionPostNode tests  to verify the change

(or)

If your pull request contain incompatible change, you should also add it to `docs/docs/en/guide/upgrede/incompatible.md`

### CODE REVIEW

File: dolphinscheduler-service/src/main/java/org/apache/dolphinscheduler/service/utils/DagHelper.java [hunk 1]
```diff
@@ -384,13 +384,40 @@ public static List<Long> parseConditionTask(Long nodeCode,
...
+     * @param conditionTaskList
+     * @param skipTaskAfterNode
+     */
+    private static void parseConditionNodeAfterSkippedNode(Map<Long, TaskNode> skipTaskNodeList,
+                                                           DAG<Long, TaskNode, TaskNodeRelation> dag,
+                                                           List<Long> conditionTaskList,
+                                                           Map<Long, TaskNode> skipTaskAfterNode) {
+
+        for (Long taskCode : skipTaskAfterNode.keySet()) {
+            Set<Long> subsequentNodes = dag.getSubsequentNodes(taskCode);
```
Comment (fuchanghai) [inline dolphinscheduler-service/src/main/java/org/apache/dolphinscheduler/service/utils/DagHelper.java:411]: I am puzzled by this, the ```skipTaskAfterNode ```parameters are all nodes not included in the ```dag```, so can you still get them through the ```dag```?

Comment (fuchanghai) [reply dolphinscheduler-service/src/main/java/org/apache/dolphinscheduler/service/utils/DagHelper.java:411]: I know what's going on, ignore the above question

File: dolphinscheduler-service/src/main/java/org/apache/dolphinscheduler/service/utils/DagHelper.java [hunk 2]
```diff
@@ -439,22 +466,26 @@ private static List<Long> skipTaskNode4Switch(TaskNode taskNode,
...
-            return;
+            return skipNodesInThisMethod;
         }
         skipTaskNodeList.putIfAbsent(skipNodeCode, dag.getNode(skipNodeCode));
+        skipNodesInThisMethod.put(skipNodeCode, dag.getNode(skipNodeCode));
         Collection<Long> postNodeList = dag.getSubsequentNodes(skipNodeCode);
         for (Long post : postNodeList) {
             TaskNode postNode = dag.getNode(post);
             if (isTaskNodeNeedSkip(postNode, skipTaskNodeList)) {
-                setTaskNodeSkip(post, dag, completeTaskList, skipTaskNodeList);
```
Comment (fuchanghai) [inline dolphinscheduler-service/src/main/java/org/apache/dolphinscheduler/service/utils/DagHelper.java:455]: If the recursion here is removed, the logic will only work on the first node

**Review by fuchanghai — CHANGES_REQUESTED** [review]

**Comment by fuchanghai** [comment]
从我的观点来看，首先c1节点的某条分支是非执行时，并且这条非执行分支上有condition 节点（以下成为c2），并且这个c2节点是或的关系。我会做以下操作
1.当或的条件没有全部判断完毕，我不会将c2 放入skip 节点
2.在c2 节点信息上加入c2所依赖的各个dependNode的状态，例如c2节点依赖c1节点的那条分支是不执行的，我就设该dependNode 为false
3.当c2 所依赖的dependNode全为false时，放入skipNodeList


```
From my point of view, first of all, a branch of the c1 node is non-executable, and there is a condition node (hereinafter referred to as c2) on this non-executive branch, and this c2 node is an or relationship. I would do the following
1. I will not put c2 into the skip node when all the conditions of or have not been judged
2. Add the status of each dependNode that c2 depends on to the c2 node information. For example, the branch where c2 node depends on c1 node is not executed, so I set the dependNode to false
3. When the dependNodes that c2 depends on are all false, put them into skipNodeList
```

hi @alei1206 WDYT ? cc @Radeity

**Comment by shilei-ly** [comment]
> 从我的观点来看，首先c1节点的某条分支是非执行时，并且这条非执行分支上有condition 节点（以下成为c2），并且这个c2节点是或的关系。我会做以下操作 1.当或的条件没有全部判断完毕，我不会将c2 放入skip 节点 2.在c2 节点信息上加入c2所依赖的各个dependNode的状态，例如c2节点依赖c1节点的那条分支是不执行的，我就设该dependNode 为false 3.当c2 所依赖的dependNode全为false时，放入skipNodeList
> 
> ```
> From my point of view, first of all, a branch of the c1 node is non-executable, and there is a condition node (hereinafter referred to as c2) on this non-executive branch, and this c2 node is an or relationship. I would do the following
> 1. I will not put c2 into the skip node when all the conditions of or have not been judged
> 2. Add the status of each dependNode that c2 depends on to the c2 node information. For example, the branch where c2 node depends on c1 node is not executed, so I set the dependNode to false
> 3. When the dependNodes that c2 depends on are all false, put them into skipNodeList
> ```
> 
> hi @alei1206 WDYT ? cc @Radeity
image

@fuchanghai  你好，这部分我有其他看法，我认为这个问题的原因是特殊场景下c2无法被提交执行，而不是c2是否加入到skipNodeList的问题。如图（s1是sleep 30s， s7是sleep 10s），在此场景下，c2是不会被放入skipNodeList。当s7执行结束时，由于s1处于running状态，所以c2的dependNodes未完成，不会被提交执行。当s1完成后，会将s4放入到skipNodeList，DolphinScheduler只会检查c2节点是否需要被放入skipNodeList，而不会检查是否需要执行，造成c2无法被提交执行。我所做的是获取c1之后所有被跳过的节点，然后判断被跳过节点后面是否存在Condition类型任务（如c2）。如果有则将其加入到队列并尝试提交执行。如果c2不满足执行条件，则在后续逻辑中会被移除队列。


@fuchanghai Hi, In my opinion, I think the reason for this problem is that c2 cannot be committed for execution in special scenarios, rather than whether c2 is added to skipNodeList. As shown in the figure (s1 is sleep 30s, s7 is sleep 10s), in this scenario, c2 is not put into the skipNodeList. When s7 finish, c2's dependNodes are not completed and c2 will not be submitted for execution because s1 is in the running state. When s1 is completed, s4 is added to skipNodeList, while DolphinScheduler only checks whether c2 needs to be added to skipNodeList, but not check whether c2 needs to be executed. My job is to get all the skipped nodes after c1, and then determine if there are any Condition type tasks (such as c2) after the skipped nodes. If so, add it to the queue and try to commit for execution. If c2 does not meet the execution condition, it is removed from the queue in subsequent logic.

**Pull request closed without merging**